unit_tests/filter_test.cc

```diff
@@ -0,0 +1,197 @@
+
+    BPFilter f(2, sampling_rate, centre, width, in, out);
+
+    // High freq sin
+    const unsigned int low_freq = 30;
+    Audio low_sin_audio;
+
+    for (int i = 0; i < sample_size; i++) {
+        low_sin_audio.push_back(100 * (sin(2 * i * pi * low_freq) + 1));
+    }
+
+    Audio low_out = f.filter(low_sin_audio);
+
+    for (int i = 0; i < sample_size; i++) {
+        // std::cerr << "Low frequency at amplitude: " << low_out[i] << std::endl;
+        EXPECT_TRUE(low_out[i] < threshold) << "Low frequency not attentuated";
+    }
+}
+
+// Test filter function runs in less than 0.5ms
+TEST(FilterTest, RunTimeTest) {
+    const unsigned int centre = 75;
+    const unsigned int width = 20;
+    const unsigned int max_delay = 500;
+    Pipe<Audio> in;
+    Pipe<Audio> out;
+
+    BPFilter f(2, sampling_rate, centre, width, in, out);
+
+    // High freq sin
+    const unsigned int freq = 30;
+    Audio sin_audio;
+
+    for (int i = 0; i < sample_size; i++) {
+        sin_audio.push_back(100 * (sin(2 * i * pi * freq) + 1));
+    }
+    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
+
+    f.filter(sin_audio);
+
+    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
+
+    auto diff = std::chrono::duration_cast<std::chrono::microseconds>(end - begin);
+    EXPECT_TRUE(diff.count() < max_delay) << "Time difference = " << diff.count() << "[µs]" << std::endl;
+}
+
+TEST(FilterTest, ThreadAndMessaging) {
+    const unsigned int centre = 75;
+    const unsigned int width = 20;
+
+    Pipe<Audio> in_pipe;
+    Pipe<Audio> out_pipe;
+
+    BPFilter f(2, sampling_rate, centre, width, in_pipe, out_pipe);
+
+    f.run();
+
+    std::thread input_thread{[&in_pipe] {
+        // Pass freq sin
+        const unsigned int pass_freq = 75;
+        Audio pass_sin_audio;
+
+        for (int i = 0; i < sample_size; i++) {
+            pass_sin_audio.push_back(100 * (sin(2 * i * pi * pass_freq) + 1));
+        }
+
+        auto *in = &in_pipe;
+        {
+            std::lock_guard<std::mutex> lk(in->cond_m);
+            in->queue.push(pass_sin_audio);
+        }
+        in->cond.notify_all();
+    }};
+
+    // listen for output
+    std::thread output_thread{[&out_pipe] {
+        std::unique_lock<std::mutex> lk(out_pipe.cond_m);
+        out_pipe.cond.wait(lk, [&out_pipe] { return out_pipe.queue.empty() == false; });
+
+        Audio pass_out = out_pipe.queue.front();
+        unsigned short max = 0;
+        for (int i = 0; i < sample_size; i++) {
+            if (pass_out[i] > max) {
+                max = pass_out[i];
+            }
+        }
+        EXPECT_TRUE(max > threshold) << "Pass frequency attentuated";
+    }};
+
+    std::cerr << "Exiting" << std::endl;
+    f.stop();
+    std::cerr << "Filter stopped" << std::endl;
+    input_thread.join();
+    std::cerr << "Input stopped" << std::endl;
+    output_thread.join();
+    std::cerr << "Output stopped" << std::endl;
+}
```

Comment:
I'm not convinced this unit test doesn't suffer from the same hanging problem mine did. Can you add synchronisation logic to ensure that the filter will only every try to loop once.